### PR TITLE
[Blog Post] Tracing SSL/TLS connections using eBPF.

### PR DESCRIPTION
**Pull request by htroisi**

### CODE REVIEW

File: content/blog/ebpf-openssl-tracing/index.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+This post is part of our ongoing series sharing how you can use eBPF to debug applications without recompilation or redeployment. For more eBPF projects, check out:
+
+- [Part 1](https://blog.pixielabs.ai/blog/ebpf-function-tracing/post/): Debugging Go function arguments in prod
+- [Part 2](https://blog.pixielabs.ai/ebpf-http-tracing/): Full body HTTP request/response tracing
+
+## Tracing with BPF
+
+[BPF](https://www.brendangregg.com/ebpf.html) can be used to interrupt the execution of a program at virtually any point—much like how one can set a breakpoint in an application using a debugger.
+
+However, unlike a debugger, which stops a program indefinitely to allow you to poke around, with BPF the kernel simply runs a BPF program when the specified trigger occurs and then immediately resumes execution of the original program.
```
Comment (oazizi000) [inline content/blog/ebpf-openssl-tracing/index.md:23]: "with BPF,"

File: content/blog/ebpf-openssl-tracing/index.md [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+This strategy is well supported by BPF *uprobes*. With uprobes, we set our triggers to be events that happen in user-space. Often, uprobes are set in the user’s own compiled code, but there’s nothing preventing us from placing them on a shared library.
+
+By placing the uprobe on a *shared* library, we end up tracing *all* applications that use the shared library, so we’ll need to filter out the data for the processes we’re actually interested in. This is no different than when we put a kprobe on Linux’s `send()` and `recv()` syscalls, though. Those kprobes also end up tracing *all* applications —- in fact, they trace more applications, as not all applications will use OpenSSL, but all applications will go through the kernel. In the context of a full system tracer, however, a probe on *shared* library is actually an advantage, since a single probe gives us wide observability coverage.
```
Comment (oazizi000) [inline content/blog/ebpf-openssl-tracing/index.md:64]: not all applications will use OpenSSL -> not all applications use OpenSSL
all applications will go through the kernel -> all aplications go through the kernel

File: content/blog/ebpf-openssl-tracing/index.md [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+Compiler optimizations like inlining can often mean that the symbol you’re looking for is not available, so this is something to watch out for. Nevertheless, as long as the symbol is available in the binary, you can trace it.
+
+Now, what if you want to attach a uprobe to a shared library? In this case, we’re dealing with the `.so` shared library as our object file instead of the program we’ve compiled, since the shared library is where the code of interest lives. To start, let’s see if we can find the `SSL_write` or `SSL_read` symbol that we want to probe. To do this, we run `nm` directly on the `libssl.so` file.
+
+```bash
+$ nm /usr/lib/x86_64-linux-gnu/libssl.so.1.1
+nm: /usr/lib/x86_64-linux-gnu/libssl.so.1.1: no symbols
+```
+
+Uh-oh. No symbols? They’ve been stripped. So how will we call `attach_uprobe`?
```
Comment (oazizi000) [inline content/blog/ebpf-openssl-tracing/index.md:95]: They've been stripped? (add question mark)

File: content/blog/ebpf-openssl-tracing/index.md [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+
+## Putting it all together
+
+A fully working version of the OpenSSL tracer can be found [here](https://github.com/pixie-io/pixie-demos/tree/main/openssl-tracer). To run the demo, follow the directions in the [`README`](https://github.com/pixie-io/pixie-demos/blob/main/openssl-tracer/README.md) file.
+
+The main project files are:
+
+- `Openssl_tracer.cc`: the user-space tracer
+- `Openssl_tracer_bpf_funcs.c`: the BPF probes to deploy
+- `probe_deployment.h/cc`: a thin wrapper around BCC
```
Comment (oazizi000) [inline content/blog/ebpf-openssl-tracing/index.md:128]: decapitcalize the file names.

**Review by oazizi000 — APPROVED** [review]

**Review by aimichelle — APPROVED** [review]

**Pull request merged**